knowhere/feder/HNSW.h

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#pragma once
+
+#include <memory>
+#include <tuple>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+
+#include "knowhere/feder/IDSet.h"
+
+namespace knowhere::feder::hnsw {
+
+///////////////////////////////////////////////////////////////////////////////
+// index view
+struct NodeInfo {
+    int64_t id_;
+    std::vector<int64_t> links_;
+
+    NodeInfo() = default;
+
+    NodeInfo(const int64_t id, const std::vector<int64_t>& links) = delete;
+
+    NodeInfo(const int64_t id, std::vector<int64_t>&& links) : id_(id), links_(links) {
+    }
+
+    NodeInfo(const NodeInfo& other) {
+        id_ = other.id_;
+        links_ = other.links_;
+    }
+
+    NodeInfo(NodeInfo&& other) {
```

Comment:
A small trick here, move constructor/operator have to be marked as `noexcept`. Since it is not revertable, which means if an exception occur in move, the half-moved things cannot be reverted like a copy.

Reply:
But as mentioned before, we may not need this since its perform is be-default.

Reply:
change it to default